### PR TITLE
Add roadmap

### PR DESCRIPTION
**Pull request by jarrodmillman**
We need a roadmap for the  NumFOCUS application Ross is preparing.  I have been meaning to draft this for a while, but this is good motivation.

It is based in part on things we've been discussing as well as some things from the CZI proposal.  I am hoping that at least @dschult and @rossbar can carefully review this before Thursday.  We can discuss it in more depth on Thursday.  I would like to merge this in about a week, so it would be good to have a few iterations on it before then.

I think it is about the right length and should perhaps be shortened a bit.  Please feel free to remove things you aren't interested in.  If I missed anything major that we want, please feel free to push changes to my branch.

It is a living document, so we will add and remove stuff later as desired.

### CODE REVIEW

File: doc/developer/roadmap.rst [hunk 1]
```diff
@@ -2,3 +2,102 @@
...
+
+Performance
+-----------
+
+Speed improvements, lower memory usage, and the ability to parallelize
+algorithms are beneficial to most science domains and use cases.
+
+This may include:
+
+- investigate improved integration with NumPy, SciPy, and Pandas
```
Comment (ericmjl) [inline doc/developer/roadmap.rst:46]: Would it help to make the broader point as well as the specific examples?

```suggestion
- investigate improved integration with the numerical computing stack in Python, e.g. NumPy, SciPy and Pandas
```

Just a thought, not a big deal.

File: doc/developer/roadmap.rst [hunk 2]
```diff
@@ -2,3 +2,102 @@
...
+
+This includes:
+
+- improving continuous integration
+- making code base more approachable
+- creating new pathways beyond volunteer effort
+- growing maintainers and leadership
+- increasing diversity of developer community
+
+Performance
```
Comment (ericmjl) [inline doc/developer/roadmap.rst:38]: On performance, cuGraph now accepts NetworkX graph objects too! :)

https://twitter.com/datametrician/status/1312377501423992832

**Review by ericmjl — APPROVED** [review]
Thanks for taking the lead on this, @jarrodmillman!

File: doc/developer/roadmap.rst (hunk 2 above)
Comment (jarrodmillman) [reply doc/developer/roadmap.rst:38]: Nice!

Comment (jarrodmillman) [reply doc/developer/roadmap.rst:38]: We should add a gallery example using cuGraph or maybe a new directory with several cuGraph examples.

File: doc/developer/roadmap.rst [hunk 3]
```diff
@@ -2,3 +2,102 @@
...
+This may include:
+
+- investigate improved integration with NumPy, SciPy, and Pandas
+- investigate use of Cython and/or Numba
+- investigate parallelization infrastructure
+
+This may involve implementing a benchmarking system using something
+like airspeed velocity (https://asv.readthedocs.io/en/stable/).
+It would also include review existing comparisons between NetworkX
+and other packages
```
Comment (rossbar) [inline doc/developer/roadmap.rst:53]:
```suggestion
and other packages.
```

File: doc/developer/roadmap.rst [hunk 4]
```diff
@@ -2,3 +2,102 @@
...
+
+Documentation
+-------------
+
+We’d like to improve the content, structure, and presentation of the NetworkX
+documentation. Some specific goals include:
+
+- more elaborate gallery examples
+- more domain-specific documentation (e.g., NetworkX for Geneticists, NetworkX for Neuroscientists)
```
Comment (rossbar) [inline doc/developer/roadmap.rst:62]:
```suggestion
- more domain-specific documentation (e.g., NetworkX for Geneticists, NetworkX for Neuroscientists, etc.)
```

File: doc/developer/roadmap.rst [hunk 5]
```diff
@@ -2,3 +2,102 @@
...
+
+Interoperability
+----------------
+
+We'd like to improve interoperability with the rest of the scientific Python
+ecosystem.
+This includes projects we depend on (e.g., NumPy, SciPy, Pandas, Matplotlib)
+as well as ones we don't (e.g., Geopandas).
+We would also like to be able to seamlessly exchange graphs with other
+NetworkX analysis software.
```
Comment (rossbar) [inline doc/developer/roadmap.rst:89]:
```suggestion
network analysis software.
```

**Review by rossbar — APPROVED** [review]
Thanks for putting this together @jarrodmillman . I think it provides really nice high-level coverage of project goals. There are certainly opportunities for expansion, but the individual sections can be revisited as more concrete investigation (e.g. in the form of informational NXEP's) becomes available.

File: doc/developer/roadmap.rst [hunk 6]
```diff
@@ -2,3 +2,104 @@
...
+Interoperability
+----------------
+
+We'd like to improve interoperability with the rest of the scientific Python
+ecosystem.
+This includes projects we depend on (e.g., NumPy, SciPy, Pandas, Matplotlib)
+as well as ones we don't (e.g., Geopandas).
+We would also like to be able to seamlessly exchange graphs with other
+network analysis software.
+
```
Comment (dschult) [inline doc/developer/roadmap.rst:92]: Another way to integrate with other scientific python ecosystem tools is to take on features from the other tools that are useful.  And we should develop tools to ease use of NetworkX from within these other tools. We have discussed the following. Maybe they should be included in the roadmap?

```suggestion

Examples of improvements of interoperability include providing a more pandas-like interface for the ```__getitem__``` dunder function of node and edge views (NXEP2). Also developing a universal method to represent a graph as a single sequence of ```nodes_and_edges``` objects that allow attribute dicts, nodes and edges as discussed for graph generators in #3036 

```

**Review by dschult — APPROVED** [review]
I think this is a really useful document to have in the repo and build/change/look at for ideas about where we are headed. It could be quite useful when reviewing NXEPs as well.

**Comment by jarrodmillman** [comment]
@dschult Would you take a closer look at the ``Algorithms`` section?  It needs to be rewritten.  Currently, if is just a copy and paste from the CZI grant milestones.  It would be great if you (or anyone else who is motivated) could suggest some text.

I was thinking it should say something about the types of algorithms we would be most excited about.  Or discuss how active graph algorithms research is.  Do we want to emphasize local graph algorithms?  Heuristics?  Quasilinear?  Is there a type of algorithm that we are completely missing?

I may be missing it, but I don't think we list anywhere what types of algorithms we are most interested in including.  Maybe it should also include something about performance vs. readability.  For example, we like having pure Python implementations even when we also provide faster versions.  Could we add a sentence here stating "our policy/preference" on this?  Maybe we should also link to our values statement (https://networkx.github.io/documentation/latest/developer/values.html)?

File: doc/developer/roadmap.rst [hunk 7]
```diff
@@ -2,3 +2,115 @@
...
+```nodes_and_edges``` objects that allow attribute dicts, nodes and edges as
+`discussed for graph generators
+<https://github.com/networkx/networkx/issues/3036>`_.
+
+Visualization
+-------------
+
+Visualization is not a focus on NetworkX, but it is a major feature for
+many users.
+We need to enhance the drawing tools for NetworkX.
```
Comment (stefanv) [inline doc/developer/roadmap.rst:109]: Unclear whether this means "support external projects to do so" or "build these tools in-house".

File: doc/developer/roadmap.rst [hunk 8]
```diff
@@ -2,3 +2,115 @@
...
+Sustainability
+--------------
+
+We aim to reduce barriers to contribution, speed up pull request (PR) review,
+onboard new maintainers, and attract new developers to ensure the long-term
+sustainability of NetworkX.
+
+This includes:
+
+- improving continuous integration
```
Comment (stefanv) [inline doc/developer/roadmap.rst:32]: Not sure if it belong in this document, but something about being present at workshops at conferences / doing outreach and "marketing" may be relevant here.

Comment (stefanv) [reply doc/developer/roadmap.rst:32]: Establish a good venue for users to ask questions and get help?

**Review by stefanv — APPROVED** [review]

File: doc/developer/roadmap.rst [hunk 9]
```diff
@@ -2,3 +2,115 @@
...
+But, we need to decide on an API and perhaps implement some
+helper code to make it consistent.
+
+Documentation
+-------------
+
+We’d like to improve the content, structure, and presentation of the NetworkX
+documentation. Some specific goals include:
+
+- longer gallery examples
```
Comment (stefanv) [inline doc/developer/roadmap.rst:62]: Perhaps, specifically, ones that use *real world data*

**Review by MridulS — APPROVED** [review]

File: doc/developer/roadmap.rst (hunk 7 above)
Comment (jarrodmillman) [reply doc/developer/roadmap.rst:113]: I am not sure what the answer to that question.  We will need to figure that out as part of the roadmap.

File: doc/developer/roadmap.rst (hunk 8 above)
Comment (jarrodmillman) [reply doc/developer/roadmap.rst:32]: I agree, but not sure it belongs here either.

**Pull request merged**